File: internal/tracing/status.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package tracing

import (
	"strings"
	"sync"
)

const (
	StatusOk         = "ok"
	StatusError      = "error"
	StatusInProgress = "in_progress"
)

var errorMarkers = []string{"failed", "error", "rejected"}

func EventStatus(eventType string) string {
	lower := strings.ToLower(eventType)
	for _, marker := range errorMarkers {
		if strings.Contains(lower, marker) {
			return StatusError
		}
	}
	return StatusOk
}

var (
	terminalMarkersMu sync.RWMutex
	terminalMarkers   = map[string][]string{}
)

// RegisterTerminalMarkers declares the event names that close out a trace for
// the given source. Connectors call this at boot so the core stays unaware of
// connector-specific event names.
func RegisterTerminalMarkers(source string, markers ...string) {
	terminalMarkersMu.Lock()
	defer terminalMarkersMu.Unlock()
	terminalMarkers[source] = append(terminalMarkers[source], markers...)
}

type TraceEventStatus struct {
	Source    string
	EventType string
}

func setTerminalMarkersForTest(t interface{ Cleanup(func()) }, source string, markers []string) {
	terminalMarkersMu.Lock()
	prev, hadPrev := terminalMarkers[source]
	terminalMarkers[source] = markers
	terminalMarkersMu.Unlock()
	t.Cleanup(func() {
		terminalMarkersMu.Lock()
		defer terminalMarkersMu.Unlock()
		if hadPrev {
			terminalMarkers[source] = prev
		} else {
			delete(terminalMarkers, source)
		}
	})
}

// TraceStatus is "error" if any event errored, "ok" if a registered terminal
// marker is present, else "in_progress".
func TraceStatus(events []TraceEventStatus) string {
	terminalMarkersMu.RLock()
	snapshot := make(map[string][]string, len(terminalMarkers))
	for k, v := range terminalMarkers {
		snapshot[k] = v
	}
	terminalMarkersMu.RUnlock()

	hasError := false
	hasTerminal := false
	for _, e := range events {
		if EventStatus(e.EventType) == StatusError {
			hasError = true
		}
		for _, m := range snapshot[e.Source] {
			if e.EventType == m {
				hasTerminal = true
			}
		}
	}
	if hasError {
		return StatusError
	}
	if hasTerminal {
		return StatusOk
	}
	return StatusInProgress
}

File: web/src/features/core/signer/status.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import type { StatusBadgeVariant } from "@/shared/ui/StatusBadge";
import type { Certificate } from "./schemas";

export type CertificateStatus = "active" | "expired" | "revoked";

export const CERTIFICATE_STATUS_LABELS: Record<CertificateStatus, string> = {
  active: "Active",
  expired: "Expired",
  revoked: "Revoked",
};

// Revocation takes precedence over time-based validity.
export function getCertificateStatus(cert: Certificate, now: Date = new Date()): CertificateStatus {
  if (cert.revoked) return "revoked";
  const nowSeconds = Math.floor(now.getTime() / 1000);
  if (cert.valid_before < nowSeconds) return "expired";
  return "active";
}

export function statusBadgeVariantFromCertificateStatus(
  status: CertificateStatus,
): StatusBadgeVariant {
  if (status === "active") return "active";
  if (status === "expired") return "expired";
  return "deleted";
}

export function formatUnixSeconds(seconds: number): string {
  return new Date(seconds * 1000).toLocaleString();
}

export function formatUnixDate(seconds: number): string {
  return new Date(seconds * 1000).toLocaleDateString(undefined, {
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}

export function formatCertificateLifetime(cert: Certificate): string {
  const seconds = Math.max(0, cert.valid_before - cert.valid_after);
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  if (hours > 0 && minutes === 0) return `${hours} hour${hours === 1 ? "" : "s"}`;
  if (hours > 0) return `${hours}h ${minutes}m`;
  if (minutes > 0) return `${minutes} minute${minutes === 1 ? "" : "s"}`;
  return `${seconds} seconds`;
}
